src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java

```diff
@@ -32,43 +37,126 @@
         super(variableResolver, indentOptions);
         this.template = template;
-        this.existingVariables = new ArrayList<>();
+        this.useTemplateForInvocationOnlySnippet = useTemplateForInvocationOnlySnippet;
     }
 
     @Override
     public void startSnippet() {
 
     }
 
+    @SuppressWarnings("DataFlowIssue")
     @Override
     public void endSnippet() {
+        List<SnippetTemplateSegment> effectiveTemplateSegments = templateSegments;
+
+        // If we should not use a template for an invocation-only snippet, see whether this fits the pattern
+        if (!useTemplateForInvocationOnlySnippet) {
+            boolean hasInvocation = false;
+            boolean inInvocation = false;
+            boolean hasVariablesOutsideInvocation = false;
+            List<SnippetTemplateSegment> simplifiedTemplateSegments = new ArrayList<>(templateSegments.size());
+            for (SnippetTemplateSegment templateSegment : templateSegments) {
+                // Keep track of invocations
+                if (templateSegment.isTextSegment()) {
+                    String textSegment = templateSegment.getTextSegment();
+                    if (!inInvocation && isInvocationStart(textSegment)) {
+                        // Mark us as having an invocation and being in it now
+                        hasInvocation = true;
+                        inInvocation = true;
+
+                        // Add the invocation start text segment and an end variable inside the invocation
+                        simplifiedTemplateSegments.add(templateSegment);
+                        simplifiedTemplateSegments.add(SnippetTemplateSegment.endVariable());
+                    } else if (inInvocation && isInvocationEnd(textSegment)) {
+                        // Mark us as no longer being in an invocation
+                        inInvocation = false;
+                        // Add the invocation end text segment
+                        simplifiedTemplateSegments.add(templateSegment);
+                    }
+                    // Otherwise only add text segments outside the invocation as the invocation should only contain the end variable
+                    else if (!inInvocation) {
+                        simplifiedTemplateSegments.add(templateSegment);
+                    }
+                }
+                // If we find a variable not in an invocation, we can't simplify this template
+                else if ((templateSegment.isVariable() || templateSegment.isNamedVariable()) && !inInvocation) {
+                    hasVariablesOutsideInvocation = true;
+                    break;
+                }
+            }
 
+            // If we found an invocation and all variables were inside of it, use the simplified template segments
+            if (hasInvocation && !hasVariablesOutsideInvocation) {
+                effectiveTemplateSegments = simplifiedTemplateSegments;

```

Comment:
If it looks like an invocation-only snippet, use the simplified segments below when building the actual template.